x-pack/plugins/security_solution/scripts/endpoint/common/fleet_server/fleet_server_services.ts

```diff
@@ -0,0 +1,673 @@
+import type { AgentPolicy, CreateAgentPolicyResponse, Output } from '@kbn/fleet-plugin/common';
+import {
+  AGENT_POLICY_API_ROUTES,
+  API_VERSIONS,
+  FLEET_SERVER_PACKAGE,
+  FLEET_SERVER_SERVERS_INDEX,
+  PACKAGE_POLICY_SAVED_OBJECT_TYPE,
+} from '@kbn/fleet-plugin/common';
+import type {
+  FleetServerHost,
+  GetOneOutputResponse,
+  PutOutputRequest,
+} from '@kbn/fleet-plugin/common/types';
+import type {
+  PostFleetServerHostsRequest,
+  PostFleetServerHostsResponse,
+} from '@kbn/fleet-plugin/common/types/rest_spec/fleet_server_hosts';
+import {
+  fleetServerHostsRoutesService,
+  outputRoutesService,
+} from '@kbn/fleet-plugin/common/services';
+import axios from 'axios';
+import * as https from 'https';
+import {
+  CA_TRUSTED_FINGERPRINT,
+  FLEET_SERVER_CERT_PATH,
+  FLEET_SERVER_KEY_PATH,
+  fleetServerDevServiceAccount,
+} from '@kbn/dev-utils';
+import { maybeCreateDockerNetwork, SERVERLESS_NODES, verifyDockerInstalled } from '@kbn/es';
+import { resolve } from 'path';
+import type * as estypes from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import {
+  RETRYABLE_TRANSIENT_ERRORS,
+  retryOnError,
+} from '../../../../common/endpoint/data_loaders/utils';
+import { isServerlessKibanaFlavor } from '../stack_services';
+import type { FormattedAxiosError } from '../format_axios_error';
+import { catchAxiosErrorFormatAndThrow } from '../format_axios_error';
+import {
+  fetchFleetOutputs,
+  fetchFleetServerHostList,
+  fetchFleetServerUrl,
+  fetchIntegrationPolicyList,
+  generateFleetServiceToken,
+  getAgentVersionMatchingCurrentStack,
+  getFleetElasticsearchOutputHost,
+  waitForHostToEnroll,
+} from '../fleet_services';
+import { dump } from '../../endpoint_agent_runner/utils';
+import { getLocalhostRealIp } from '../localhost_services';
+import { isLocalhost } from '../is_localhost';
+
+export const FLEET_SERVER_CUSTOM_CONFIG = resolve(__dirname, './fleet_server.yml');
+
+interface StartedServer {
+  /** The type of virtualization used to start the server */
+  type: 'docker';
+  /** The ID of the server */
+  id: string;
+  /** The name of the server */
+  name: string;
+  /** The url (including port) to the server */
+  url: string;
+  /** Stop server */
+  stop: () => Promise<void>;
+  /** Any information about the server */
+  info?: string;
+}
+
+interface StartFleetServerOptions {
+  kbnClient: KbnClient;
+  logger: ToolingLog;
+  /** Policy ID that should be used to enroll the fleet-server agent */
+  policy?: string;
+  /** Agent version */
+  version?: string;
+  /** Will start fleet-server even if its detected that it is already running */
+  force?: boolean;
+  /** The port number that will be used by fleet-server to listen for requests */
+  port?: number;
+}
+
+interface StartedFleetServer extends StartedServer {
+  /** The policy id that the fleet-server agent is running with */
+  policyId: string;
+}
+
+export const startFleetServer = async ({
+  kbnClient,
+  logger,
+  policy,
```

Comment:
correct - if `policy` (Fleet agent policy UUID) is provided, then we don't create one automatically if one does not yet exist. It could be used from CI or CLI - it does not matter. It just provides the consumer the ability to define the specific agent policy ID to use in starting the server